src/disk_cache.rs

```diff
@@ -0,0 +1,955 @@
+                .get_opts(
+                    &location,
+                    GetOptions {
+                        range: Some(GetRange::Bounded(range)),
+                        ..Default::default()
+                    },
+                )
+                .await?;
+
+            // save it to the disk cache, we'll ignore the error on writing to disk here, just return
+            // the bytes from the object store.
+            let meta = get_result.meta.clone();
+            let bytes = get_result.bytes().await?;
+            entry.save_meta(&meta).await.ok();
+            entry.save_part(part_id, &bytes.clone()).await.ok();
+            Ok(bytes.slice(range_in_part))
+        })
+    }
+
+    // given the range and object size, return the canonicalized `Range<usize>` with concrete start and
+    // end.
+    fn canonicalize_range(
+        &self,
+        range: Option<GetRange>,
+        object_size: usize,
+    ) -> object_store::Result<Range<usize>> {
+        let (start_offset, end_offset) = match range {
+            None => (0, object_size),
+            Some(range) => match range {
+                GetRange::Bounded(range) => {
+                    if range.start >= object_size {
+                        return Err(object_store::Error::Generic {
+                            store: "cached_object_store",
+                            source: Box::new(InvalidGetRange::StartTooLarge {
+                                requested: range.start,
+                                length: object_size,
+                            }),
+                        });
+                    }
+                    if range.start >= range.end {
+                        return Err(object_store::Error::Generic {
+                            store: "cached_object_store",
+                            source: Box::new(InvalidGetRange::Inconsistent {
+                                start: range.start,
+                                end: range.end,
+                            }),
+                        });
+                    }
+                    (range.start, range.end.min(object_size))
+                }
+                GetRange::Offset(offset) => {
+                    if offset >= object_size {
+                        return Err(object_store::Error::Generic {
+                            store: "cached_object_store",
+                            source: Box::new(InvalidGetRange::StartTooLarge {
+                                requested: offset,
+                                length: object_size,
+                            }),
+                        });
+                    }
+                    (offset, object_size)
+                }
+                GetRange::Suffix(suffix) => (object_size.saturating_sub(suffix), object_size),
+            },
+        };
+        Ok(Range {
+            start: start_offset,
+            end: end_offset,
+        })
+    }
+
+    fn align_get_range(&self, range: &GetRange) -> GetRange {
```

Comment:
would it be simpler to canonicalize the range first and then align the canonicalized range?

Reply:
Got it, that makes sense thanks.